streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java

```diff
@@ -217,193 +232,195 @@ public void shouldUpdateOffsetsUponCompletion() {
+
     }
 
     @Test
     public void shouldThrowTaskMigratedExceptionOnUnknownProducerIdException() {
-        final KafkaException exception = new UnknownProducerIdException("KABOOM!");
-        final RecordCollector collector = new RecordCollectorImpl(
-            taskId,
-            streamsConfig,
-            logContext,
-            streamsMetrics,
-            consumer,
-            id -> new MockProducer<byte[], byte[]>(cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer) {
-                @Override
-                public synchronized Future<RecordMetadata> send(final ProducerRecord<byte[], byte[]> record, final Callback callback) {
-                    throw new KafkaException(exception);
-                }
-            }
-        );
+        final UnknownProducerIdException exception = new UnknownProducerIdException("KABOOM!");
+        // we need to mimic that `send()` always wraps error in a KafkaException
+        eosMockProducer.sendException = new KafkaException(exception);
 
-        final TaskMigratedException thrown = assertThrows(TaskMigratedException.class, () ->
-            collector.send(topic, "3", "0", null, null, stringSerializer, stringSerializer, streamPartitioner)
+        final TaskMigratedException thrown = assertThrows(
+            TaskMigratedException.class,
+            () -> eosCollector.send(topic, "3", "0", null, null, stringSerializer, stringSerializer, streamPartitioner)
         );
+
         assertEquals(exception, thrown.getCause());
+        assertThat(thrown.getMessage(), equalTo("Producer cannot send records anymore since it got fenced"));
     }
 
     @Test
     public void shouldThrowTaskMigratedExceptionOnSubsequentCallWhenProducerFencedInCallback() {
         final KafkaException exception = new ProducerFencedException("KABOOM!");
         final RecordCollector collector = new RecordCollectorImpl(
-            taskId,
-            streamsConfig,
             logContext,
-            streamsMetrics,
-            null,
-            id -> new MockProducer<byte[], byte[]>(cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer) {
-                @Override
-                public synchronized Future<RecordMetadata> send(final ProducerRecord record, final Callback callback) {
-                    callback.onCompletion(null, exception);
-                    return null;
-                }
-            }
+            taskId,
+            mockConsumer,
+            new StreamsProducer(
+                new MockProducer<byte[], byte[]>(cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer) {
+                    @Override
+                    public synchronized Future<RecordMetadata> send(final ProducerRecord<byte[], byte[]> record, final Callback callback) {
+                        callback.onCompletion(null, exception);
```

Comment:
`MockProducer` does not support the correct handling of callback exceptions, hence, we need to do it manually. (Did not think it worth to extend `MockProducer` -- let me know what you think)